lib/ansible/module_utils/hwc_utils.py

```diff
@@ -0,0 +1,316 @@
+
+    def delete(self, url, body=None):
+        try:
+            return self._adapter.delete(
+                url, json=body,
+                headers=self._headers(), raise_exc=False)
+        except getattr(requests.exceptions, 'RequestException') as inst:
+            self.module.fail_json(msg=inst.message)
+
+    def put(self, url, body=None):
+        try:
+            return self._adapter.put(
+                url, json=body,
+                headers=self._headers(), raise_exc=False)
+        except getattr(requests.exceptions, 'RequestException') as inst:
+            self.module.fail_json(msg=inst.message)
+
+    def get_service_endpoint(self, service_type):
+        if self._endpoints.get(service_type):
+            return self._endpoints.get(service_type)
+
+        e = None
+        try:
+            e = self._session.get_endpoint_data(
+                service_type=service_type,
+                region_name=self.module.params['region']
+            )
+        except getattr(requests.exceptions, 'RequestException') as inst:
+            self.module.fail_json(msg=inst.message)
+
+        if not e or e.url == "":
+            self.module.fail_json(
+                msg="Can not find the enpoint for %s" % service_type)
+
+        url = e.url
+        if url[-1] != "/":
+            url += "/"
+
+        self._endpoints[service_type] = url
+        return url
+
+    def get_project_id(self):
+        if self._project_id:
+            return self._project_id
+        try:
+            pid = self._session.get_project_id()
+            self._project_id = pid
+            return pid
+        except getattr(requests.exceptions, 'RequestException') as inst:
+            self.module.fail_json(msg=inst.message)
+
+    def _validate(self):
+        if not HAS_REQUESTS:
+            self.module.fail_json(msg="Please install the requests library")
+
+        if not HAS_THIRD_LIBRARIES:
+            self.module.fail_json(
+                msg="Please install the keystoneauth1 library")
+
+    def _credentials(self):
+        auth = v3.Password(
+            auth_url=self.module.params['identity_endpoint'],
+            password=self.module.params['password'],
+            username=self.module.params['user_name'],
+            user_domain_name=self.module.params['domain_name'],
+            project_name=self.module.params['project_name'],
+            reauthenticate=True
+        )
+
+        return session.Session(auth=auth)
+
+    def _headers(self):
+        return {
+            'User-Agent': "Huawei-Ansible-MM-%s" % self.product,
+            'Accept': 'application/json',
+        }
+
+
+class HwcModule(AnsibleModule):
+    def __init__(self, *args, **kwargs):
+        arg_spec = kwargs.setdefault('argument_spec', {})
+
+        arg_spec.update(
+            dict(
+                identity_endpoint=dict(
+                    required=True, type='str',
+                    fallback=(env_fallback, ['IDENTITY_ENDPOINT']),
```

Comment:
Please don't use such generic env vars.
```suggestion
                    fallback=(env_fallback, ['ANSIBLE_HWC_IDENTITY_ENDPOINT']),
```

Reply:
Good Point!

Reply:
Done